### PR TITLE
Apache Karaf 4 runtime

### LINKED ISSUE #256: Switch to Karaf 4.0.0
Web sockets (see #102)  will help to simply some code in the manager.
They will also improve the user experience in the web administration.

But to implement them, we need to switch to Karaf 4.x. Unfortunately, it is still a milestone. And some bugs remain. We should start to test it and provide feedback to the Karaf team.

This is important because we need web sockets to improve the web administration.

### PR DESCRIPTION
**Pull request by bourretp**
Fix #256

We use Karaf 4.0.0.M3 instead of a proper release. Upgrading to v4.0.0, when available, should be straightforward.

*Noticeable changes:*
- Version upgrades:
    - (RT) Apache Karaf 3.0.3 => 4.0.0.M3
    - (IT) Pax Exam 4.4.0 => 4.6.0
    - (IT) Pax URL Æther 2.2.0 => 2.4.1
- Removed branding bundles, replaced with inline static branding configuration `etc/branding.properties` inside the distribution archives. Karaf 4 seems to have on long-standing issue with branding bundles. Despite this is a workaround, the solution looks better than the original ;)
- List Karaf 4 features, as the `standard` super-feature is gone.
- Add some delay in integration tests. This inexplicably allows IT to pass on certain configurations. BTW, these tests are highly scheduler-dependent.

### CODE REVIEW

**Comment by vincent-zurczak** [comment]
Integration tests indeed use Thread.sleep here and there because they involve various programs (DM, agent, messaging server).

File: karaf/roboconf-karaf-dist-agent/pom.xml [hunk 1]
```diff
@@ -84,7 +78,23 @@
 				<artifactId>karaf-maven-plugin</artifactId>
 				<configuration>
 					<bootFeatures>
-						<feature>standard</feature>
+						<!-- Karaf 4.x has split the "standard" features into many sub-features -->
+						<feature>shell-compat</feature>
```
Comment (vincent-zurczak) [inline karaf/roboconf-karaf-dist-agent/pom.xml]: Do we need all of them?

**Comment by vincent-zurczak** [comment]
Code review looks good to me.
I restarted the failed build job (Oracle JDK 8) on Travis.

If it passes, let's wait our Monday meeting before merging it.

File: karaf/roboconf-karaf-dist-agent/pom.xml (hunk 1 above)
Comment (bourretp) [inline karaf/roboconf-karaf-dist-agent/pom.xml]: For sure we don't! I've just replaced the standard feature with all its content, to make things work quickly ;)
Maybe another issue would be to lighten the distributions, removing all the unnecessary features...

Comment (vincent-zurczak) [inline karaf/roboconf-karaf-dist-agent/pom.xml]: OK. Please, create a dedicated issue.

**Comment by bourretp** [comment]
Unfortunately I've missed to raise the sleep delay in the Rest/ping integration tests. I add it ASAP.

**Comment by vincent-zurczak** [comment]
Following this morning's meeting, we will wait for the relase to be complete before merging it. Indeed, we will not have time to implement any web socket. So, it is not necessary to migrate to Karaf 4 before.

**Comment by bourretp** [comment]
Apache Karaf 4.0.0.M3 uses Apache Felix framework 5.0.0. By looking at the latest news on the felix-dev mailing list, there seems to be some serious issues in felix-5.0.0, concerning the resolver (now externalized from the core framework). While not directly affected by this issue (mainly a problem with fragment bundles), I strongly suggest to wait for a newer version of Karaf (M4?/V4.0.0?), which includes a newer version of the Felix framework/resolver (release 5.0.1 should be cut very soon).

**Comment by bourretp** [comment]
Bumped to Karaf 4.0.0.
IT pass locally. Waiting for news from Mr Travis, and let's merge! :v:

**Pull request merged**